core/server/models/api-key.js

```diff
@@ -0,0 +1,63 @@
+const crypto = require('crypto');
+const ghostBookshelf = require('./base');
+const {Role} = require('./role');
+
+const ApiKey = ghostBookshelf.Model.extend({
+    tableName: 'api_keys',
+
+    defaults() {
+        // 512bit key for HS256 JWT signing
+        let secret = crypto.randomBytes(64).toString('hex');
+
+        return {
+            secret
+        };
+    },
+
+    role() {
+        return this.belongsTo('Role');
+    },
+
+    integration() {
+        return this.belongsTo('Integration');
+    },
+
+    onSaving(/* model, attrs, options */) {
+        let tasks = [];
+
+        ghostBookshelf.Model.prototype.onSaving.apply(this, arguments);
+
+        // enforce roles which are currently hardcoded
+        // - admin key = Adminstrator role
+        // - content key = no role
+        if (this.hasChanged('type')) {
+            if (this.get('type') === 'admin') {
+                tasks.setAdminRole = Role.findOne({name: 'Admin API Client'}, {columns: ['id']})
+                    .then((role) => {
+                        this.set('role_id', role.get('id'));
+                    });
+            }
+
+            if (this.get('type') === 'content') {
+                this.set('role_id', null);
```

Comment:
I have not looked at your big PR at all. One question: will we also use `models.ApiKey.edit` or only `add`?
Because setting the role_id to null only happens when you change your type. Just curious what the plan is 👍 
Maybe it's just not part of this PR :)

Reply:
`models.ApiKey.edit` will be used. I assumed `onSaving` is called for both create and update but that assumption may be coming from other ORMs 😬

Does bookshelf call `onSaving` with "hasChanged" for all the properties on `add` or does it need to be handled separately?

Reply:
So yeah, Bookshelf triggers onSaving for insert+update. You could differentiate by using `options.method === insert` or you register `onUpdating` and `onCreating`. That is up to you.

> Does bookshelf call onSaving with "hasChanged" for all the properties on add or does it need to be handled separately?

Ähm not sure i understand 🤪

Reply:
> So yeah, Bookshelf triggers onSaving for insert+update.

That's good, that's the behaviour I wanted 😄

> Ähm not sure i understand 🤪

When `onSaving` is called with `options.method === insert` will `this.hasChanged('type')` always be true if a type value has been set?

The main reason for the "hasChanged" conditional was to avoid an extra DB call every time an API Key gets saved which could be fairly often if we're storing the `lastSeenAt` value on each API call.

Reply:
> When onSaving is called with options.method === insert will this.hasChanged('type') always be true if a type value has been set?

Exceptions:

For adding: if the user has not set the type and there is also not default value in the model, `hasChanged` will be false.

For editing: if you don't edit `data.type` (previous === current value), it's false.

I hope that helps 😁

Reply:
👍 thanks! That matches what I expected. In this case the conditional being false is fine if the user hasn't set a `type` because that will throw a validation error (the type column isn't nullable and has an `isIn` validation)

Reply:
👍 Can't the user in theory set the same type as stored in the db (previous) and just a different role_id (even for the content type)?

Maybe it's not important for now 😁

Reply:
Good catch 🎣 We should set the hard coded role id if the type or role has changed. I’ll push a fix later

Reply:
Out of interest how/why would edit be used? Just for permissions, or would we allow editing the type? I think that could be dangerous - because editing the type from content to admin without updating the secret would allow a previously compromised content key to then be used with the admin api.

Reply:
From the UI perspective `edit` would only be used for generating a new secret, however this is the model layer so I think it should be more flexible. Restrictions can be placed at the API layer to limit what can be done externally.

Reply:
It's "never" right now but I imagine that could change soon when we get to implementing Members.

As a scenario, should Content API keys only ever be for public content? What if I want to pull some members-only content in to my marketing site to show what's available behind the pay-wall - should I need to use a full Admin API key for that or can I create a "private" Content API key that my site can use?

Reply:
Updated to enforce the hardcoded roles if either the type or role gets changed.

Reply:
> As a scenario, should Content API keys only ever be for public content?

Good point, as members (private) content will be on the content api

>  What if I want to pull some members-only content in to my marketing site to show what's available behind the pay-wall - should I need to use a full Admin API key for that or can I create a "private" Content API key that my site can use?

I was under the impression that we'd be making "private"("members") content api keys for that data yeah :)

The issue I have with this though, is not the different types of API keys, it's the ability to change a key from one type to another, or more specifically to change the permissions of an "active" key.

I think if you need a key with different permissions, then you should create a new key, with a new id and secret, that way any previous use of the key in a public context can be guaranteed to not be an issue.

(FWIW I don't think this would be an issue if we didn't suggest that some keys _can_ be used in the browser, if everything was serverside, then changing permissions of "active" keys wouldn't be so much of an issue.)

I don't want this convo to hold up the PR tho so i'm happy to revisit it at a later date :relaxed:

Reply:
I agree but let's revisit this later - it's working for now and we can lock it down later if necessary once we have a clearer picture of the requirements.